### ISSUE TITLE
Amplify Console Deployments Stopped Working With Deprecation of amplifyPush helper Script

### ISSUE THREAD

**Issue by gorbinphilip1** [aws-amplify/amplify-hosting#3093]
Existing deployments running on [amplify cli 5.6](https://github.com/aws-amplify/amplify-cli/tree/v5.6.0) stopped working with deprecation of amplifyPush script.

Current amplify.yml looks like this
```
    preBuild:
      commands:
        - npm install -g @aws-amplify/cli@5.6
        - node amplify-pre-build.js # some custom stuff
    build:
      commands:
        - '# Execute Amplify CLI with the helper script'
        - amplifyPush --simple # this just disappeared and Amplify console deployments doesn't deploy now
    postBuild:
      commands:
        - node amplify-post-build.js # some custom stuff
```

What's the route here to support cli 5.6 versions? 
Guessing most items mentioned here: https://docs.amplify.aws/cli/usage/headless/ won't be applicable to those old cli versions

_Originally posted by @gorbinphilip1 in https://github.com/aws-amplify/amplify-hosting/issues/3069#issuecomment-1308338433_

**Comment by gorbinphilip1** [comment]
Replicating script also in vain. Guessing envCache also got removed?
`    export STACKINFO="$(envCache --get stackInfo)"
`
Error below:
```
 - Initializing project in the cloud...
2022-11-09T09:30:03.802Z [INFO]: An error occurred when creating the CloudFormation stack
2022-11-09T09:30:03.803Z [WARNING]: ✖ Root stack creation failed
2022-11-09T09:30:03.808Z [INFO]: AlreadyExistsException: Stack [amplify-enterpriseapp-redacted-12123] already exists
```

**Comment by gorbinphilip1** [comment]
Interestingly docs still have reference to the deprecated script: https://docs.aws.amazon.com/amplify/latest/userguide/build-settings.html#branch-specific-build-settings

**Comment by gorbinphilip1** [comment]
Thanks to https://github.com/aws-amplify/amplify-hosting/issues/2096#issuecomment-888306083
Currently we are able to perform the work around with:
+ Locking nodejs to v14

+ Removing envCache dependency
amplify-push.sh
```sh
#!/usr/bin/env bash
set -e
IFS='|'

help_output () {
    echo "usage: amplify-push <--environment|-e <name>> <--simple|-s>"
    echo "  --environment  The name of the Amplify environment to use"
    echo "  --simple  Optional simple flag auto-includes stack info based on appId / env name"
    exit 1
}

init_env () {
    ENV=$1
    AMPLIFY=$2
    PROVIDERS=$3
    CODEGEN=$4
    AWSCONFIG=$5
    CATEGORIES=$6

    echo "# Start initializing Amplify environment: ${ENV}"
    if [[ -z ${STACKINFO} ]];
    then
        echo "# Initializing new Amplify environment: ${ENV} (amplify init)"
        [[ -z ${CATEGORIES} ]] && amplify init --amplify ${AMPLIFY} --providers ${PROVIDERS} --codegen ${CODEGEN} --yes || amplify init --amplify ${AMPLIFY} --providers ${PROVIDERS} --codegen ${CODEGEN} --categories ${CATEGORIES} --yes
        echo "# Environment ${ENV} details:"
        amplify env get --name ${ENV}
    else
        echo "STACKINFO="${STACKINFO}
        echo "# Importing Amplify environment: ${ENV} (amplify env import)"
        amplify env import --name ${ENV} --config "${STACKINFO}" --awsInfo ${AWSCONFIG} --yes;
        echo "# Initializing existing Amplify environment: ${ENV} (amplify init)"
        [[ -z ${CATEGORIES} ]] && amplify init --amplify ${AMPLIFY} --providers ${PROVIDERS} --codegen ${CODEGEN} --yes || amplify init --amplify ${AMPLIFY} --providers ${PROVIDERS} --codegen ${CODEGEN} --categories ${CATEGORIES} --yes
        echo "# Environment ${ENV} details:"
        amplify env get --name ${ENV}
    fi
    echo "# Done initializing Amplify environment: ${ENV}"
}

ENV=""
IS_SIMPLE=false
POSITIONAL=()
while [[ $# -gt 0 ]]
    do
    key="$1"
    case ${key} in
        -e|--environment)
        ENV=$2
        shift
        ;;
        -r|--region)
        REGION=$2
        shift
        ;;
        -s|--simple)
        IS_SIMPLE=true
        shift
        ;;
        *)
        POSITIONAL+=("$1")
        shift
        ;;
    esac
done
set -- "${POSITIONAL[@]}"

# if no provided environment name, use default env variable, then user override
if [[ ${ENV} = "" ]];
then
    ENV=${AWS_BRANCH}
fi

if [[ ${USER_BRANCH} != "" ]];
then
    ENV=${USER_BRANCH}
fi

# strip slashes, limit to 10 chars
ENV=$(echo ${ENV} | sed 's;\\;;g' | sed 's;\/;;g' | cut -c -10)

# Check valid environment name
if [[ -z ${ENV} || "${ENV}" =~ [^a-zA-Z0-9\-]+ ]] ; then help_output ; fi

AWSCONFIG="{\
\"configLevel\":\"project\",\
\"useProfile\":true,\
\"profileName\":\"default\",\
\"AmplifyAppId\":\"${AWS_APP_ID}\"\
}"
AMPLIFY="{\
\"envName\":\"${ENV}\",\
\"appId\":\"${AWS_APP_ID}\"\
}"
PROVIDERS="{\
\"awscloudformation\":${AWSCONFIG}\
}"
CODEGEN="{\
\"generateCode\":false,\
\"generateDocs\":false\
}"
CATEGORIES=""
if [[ -z ${AMPLIFY_FACEBOOK_CLIENT_ID} && -z ${AMPLIFY_GOOGLE_CLIENT_ID} && -z ${AMPLIFY_AMAZON_CLIENT_ID} ]]; then
    CATEGORIES=""
else
    AUTHCONFIG="{\
    \"facebookAppIdUserPool\":\"${AMPLIFY_FACEBOOK_CLIENT_ID}\",\
    \"facebookAppSecretUserPool\":\"${AMPLIFY_FACEBOOK_CLIENT_SECRET}\",\
    \"googleAppIdUserPool\":\"${AMPLIFY_GOOGLE_CLIENT_ID}\",\
    \"googleAppSecretUserPool\":\"${AMPLIFY_GOOGLE_CLIENT_SECRET}\",\
    \"amazonAppIdUserPool\":\"${AMPLIFY_AMAZON_CLIENT_ID}\",\
    \"amazonAppSecretUserPool\":\"${AMPLIFY_AMAZON_CLIENT_SECRET}\"\
    }"
    CATEGORIES="{\
    \"auth\":$AUTHCONFIG\
    }"
fi
# Handle old or new config file based on simple flag
if [[ ${IS_SIMPLE} ]];
then
    # https://docs.amplify.aws/cli/teams/commands#import-an-environment
    echo "# Getting Amplify CLI Cloud-Formation stack info"
    BACKEND_ENV="$(aws amplify get-backend-environment --profile=default --app-id ${AWS_APP_ID} --environment-name ${ENV} || true)"
    if [[ ${BACKEND_ENV} ]];
    then
        STACKNAME="$(echo ${BACKEND_ENV} | jq --raw-output '.backendEnvironment.stackName')"
        STACKINFO=$(aws cloudformation describe-stacks --profile=default --stack-name ${STACKNAME} | jq '.Stacks[0]')
        STACKINFO=$(echo ${STACKINFO} | jq '.Outputs | reduce .[] as $item ({}; .[$item.OutputKey] = $item.OutputValue)')
        STACKINFO="{\
            \"awscloudformation\":$STACKINFO\
        }"
    fi
    export STACKINFO
    init_env ${ENV} ${AMPLIFY} ${PROVIDERS} ${CODEGEN} ${AWSCONFIG} ${CATEGORIES}
    STACKINFO="$(amplify env get --json --name ${ENV})"
    echo "STACKINFO="${STACKINFO}
else
    # old config file, above steps performed outside of this script
    init_env ${ENV} ${AMPLIFY} ${PROVIDERS} ${CODEGEN} ${AWSCONFIG} ${CATEGORIES}
fi
```

+ Update
amplify.yml
```yml
version: 1
backend:
  phases:
    preBuild:
      commands:
        - npm install -g @aws-amplify/cli@5.6
        - yum install --assumeyes jq
        - node amplify-pre-build.js
    build:
      commands:
        - '# Execute Amplify CLI with the helper script'
        - chmod u+x ./amplify-push.sh
        - ./amplify-push.sh --simple
        - amplify push --yes
    postBuild:
      commands:
        - node amplify-post-build.js
```

**Comment by gorbinphilip1** [comment]
> Hi,
> 
> It looks like the problem is that the version of the CLI that you are relying on doesn't build with our default Node version anymore. You should be able to fix this problem by changing the Node.js version in their app's build configuration.
> 
> Following the steps in this guide: https://docs.aws.amazon.com/amplify/latest/userguide/custom-build-image.html#setup-live-updates
> 
> * Click in "Add Package Version Override"
> * Choose Node.js
> * Set 14 as their desired version.
> 
> Let me know if that fixes the problem, I'll leave this issue open for now, but I'll probably close it in a week if we don't hear back.

Nope. Overriding node.js version doesn't help. Overriding version helps to get cli 5.6 installed. Since the amplifyPush helper is deprecated & no longer doing its job.
Like I mentioned on the previous comment, we were able to get it working with the way noted there.

The true disappointment is on, earlier default build steps had a dependency on build machine envCache for pulling and deploying to the connected backend env. One morning AWS team decided to upgrade build machine, lost the envCache (not clear on how this works but pretty sure we were not able to retrieve env data from this cache) and all the builds started failing with `env already exists error` because same env exists but cache not returning that for amplify helper script to pull from.

Funny that we still find same helper script in docs even now
![Screenshot 2022-11-16 at 9 08 44 AM](https://user-images.githubusercontent.com/41995531/202077870-a991cc9b-f7b2-41b2-8401-e2569e02eaf4.png)

So to summarise, the question would be, What can we do to avoid these surprises? Apart from upgrading cli, which is it's own thread on the complications on upgrading if you have done some custom categories & cloudformation file modifications
